README.md
<h1 align="center">
  <br>
  <img src="./src/images/decipher-logo.png" alt="Decipher Logo" width="50">
  <br>
  Grey Matter UI Component Library
  <br>
</h1>

<h4 align="center">React components for Grey Matter user interfaces</h4>

<p align="center">
  <a href="https://github.com/prettier/prettier" target="_blank">
    <img src="https://img.shields.io/badge/formatted_with-prettier-ff69b4.svg"
         alt="styled with prettier">
  </a>
  <a href="https://www.styled-components.com/" target="_blank"><img src="https://img.shields.io/badge/styled_with-%F0%9F%92%85%20styled--components-orange.svg?colorB=daa357" alt="styled with styled-components"></a>
  <a href="https://circleci.com/gh/greymatter-io/gm-ui-components/tree/master" target="_blank"><img src="https://circleci.com/gh/greymatter-io/gm-ui-components/tree/master.svg?style=shield" alt="tested on CircleCI"></a>
  <a href="https://codeclimate.com/github/greymatter-io/gm-ui-components/maintainability"><img src="https://api.codeclimate.com/v1/badges/e4f79b8a8ac6216ae81c/maintainability" /></a>
  <a href="https://codeclimate.com/github/greymatter-io/gm-ui-components/test_coverage"><img src="https://api.codeclimate.com/v1/badges/e4f79b8a8ac6216ae81c/test_coverage" /></a>
  <a href="https://opensource.org/licenses/mit-license.php" target="_blank"><img src="https://badges.frapsoft.com/os/mit/mit.svg?v=103" alt="licensed under MIT"></a>
</p>

<p align="center">
  <a href="#installation">Installation</a> •
  <a href="#usage">Usage</a> •
  <a href="#documentation">Documentation</a> •
  <a href="#compatability">Compatability</a> •
  <a href="#license">License</a>
</p>

# Installation

gm-ui-components is available as an npm package. It has 4 peer dependencies:

```sh
npm install --save gm-ui-components react react-dom styled-components inter-ui react-popper @popperjs/core
```

# Usage

An example using the Button component:

```javascript
import React from 'react';
import { render } from 'react-dom';
import { Button } from 'gm-ui-components';

function App() {
  return (
    <Button
        label="Hello World!"
        type="default"
        size="xl"
        active={true}
    />
  );
}

render(<App />, document.querySelector('#app'));
```

## Theming

The component library provides 2 themes out of the box - `keen` and `keenDark`. All components use the `keen` theme by default. 

To use `keenDark` or a custom theme, wrap your app in a `ThemeProvider` and pass the theme object:

```javascript
// App.js
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from 'react-dom';
import { Button, keenDark } from 'gm-ui-components';

function Button() {
  return (
    <Button
        label="Hello World!"
        type="default"
        size="xl"
        active={true}
    />
  );
}

render(
  <ThemeProvider theme={keenDark}>
    <App />
  </ThemeProvider>, 
  document.querySelector('#app'));
```

To extend the theme, you must pass a theme object that follows our schema. Here is an example of the schema and its defaults:

```javascript
{
  name: "Keen",

  COLOR_BRAND_PRIMARY: "#00b42b",
  COLOR_BRAND_SECONDARY: "#00b42b",

  COLOR_BACKGROUND_DEFAULT: "hsl(0,0%,100%)",
  COLOR_BACKGROUND_TWO: "hsl(0,0%,97%)",
  COLOR_BACKGROUND_THREE: "hsl(0,0%,94%)",

  COLOR_CONTENT_DEFAULT: "hsla(0,0%,0%,0.85)",
  COLOR_CONTENT_CONTRAST: "hsla(0,0%,0%,1)",
  COLOR_CONTENT_MUTED: "hsla(0,0%,0%,0.6)",
  COLOR_CONTENT_NONESSENTIAL: "hsla(0,0%,0%,0.3)",

  COLOR_INTENT_HIGHLIGHT: "#00b42b",
  COLOR_INTENT_SUCCESS: "#00b42b",
  COLOR_INTENT_DANGER: "#D83D22", // WCAG AA+
  COLOR_INTENT_WARNING: "#F7CD45", // WCAG AA+
  COLOR_INTENT_INFO: "#1E6DF6", // WCAG AA+

  COLOR_KEYLINE_DEFAULT: "hsla(0,0%,0%,8%)",
  COLOR_KEYLINE_SOLID: "hsl(0,0%,8%)",

  OPACITY_FULL: "1",
  OPACITY_LIGHT: "0.7",
  OPACITY_LIGHTER: "0.5",
  OPACITY_LIGHTEST: "0.15",

  // LAYOUT
  SPACING_BASE: 8,

  ZINDEX_ABYSS: "-9999",
  ZINDEX_FLOOR: "1",
  ZINDEX_DROPDOWN: "1010",
  ZINDEX_STICKY: "1020",
  ZINDEX_FIXED: "1030",
  ZINDEX_SCRIM: "1040",
  ZINDEX_MODAL: "1050",
  ZINDEX_POPOVER: "1060",
  ZINDEX_TOOLTIP: "1070",

  CORNER_RADIUS_SHARP: "2px", // spacingScale(0.5)
  CORNER_RADIUS_INPUT: "4px", // spacingScale(1)
  CORNER_RADIUS_CARD_SM: "4px",
  CORNER_RADIUS_CARD_DEFAULT: "6px",
  CORNER_RADIUS_CARD_LG: "8px",
  CORNER_RADIUS_MAX: "90000px",

  // TYPOGRAPHY
  FONT_STACK_DEFAULT: `-apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif`,
  FONT_STACK_BRAND: `"Metropolis", "Helvetica Neue", Arial, sans-serif`,
  FONT_STACK_CODE: `"SFMono-Regular", Consolas, "Liberation Mono", Menlo, Courier, monospace`,

  FONT_SIZE_PAGE_TITLE: "40px",

  FONT_SIZE_HEADING_LG: "38px",
  FONT_SIZE_HEADING_DEFAULT: "36px",
  FONT_SIZE_HEADING_SM: "32px",

  FONT_SIZE_SUBHEADING_LG: "18px",
  FONT_SIZE_SUBHEADING_DEFAULT: "16px",
  FONT_SIZE_SUBHEADING_SM: "14px",

  FONT_SIZE_ITEM_TITLE_LG: "16px",
  FONT_SIZE_ITEM_TITLE_DEFAULT: "14px",
  FONT_SIZE_ITEM_TITLE_SM: "12px",

  FONT_SIZE_TEXT_XL: "18px",
  FONT_SIZE_TEXT_LG: "16px",
  FONT_SIZE_TEXT_DEFAULT: "14px",
  FONT_SIZE_TEXT_SM: "12px",
  FONT_SIZE_TEXT_XS: "10px",

  LINE_HEIGHT_LOOSE: 1.6,
  LINE_HEIGHT_DEFAULT: 1.4,
  LINE_HEIGHT_TIGHT: 1.2,

  FONT_WEIGHT_DEFAULT: 400,
  FONT_WEIGHT_MEDIUM: 500,
  FONT_WEIGHT_SEMIBOLD: 600,
  FONT_WEIGHT_BOLD: 700,

  LETTER_SPACING_DEFAULT: "normal"
};
```

You can extend this base theme like so:

```javascript
// ...App.js

keen = {
  ...keen,
  COLOR_BRAND_PRIMARY: "#03aaed"
}

render(
  <ThemeProvider theme={keen}>
    <App />
  </ThemeProvider>, 
  document.querySelector('#app'));
```

You can find more information on style variables [here](https://greymatter-io.github.io/gm-ui-components/?selectedKind=%20Overview%7CStyles&selectedStory=Styles&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fthemes%2Fpanel).

# Documentation

See the [Grey Matter UI Components storybook](https://greymatter-io.github.io/gm-ui-components) for more usage examples and documentation.

# Compatability

* [Full list of browser support matrix](http://browserl.ist/?q=last+2+Chrome+versions%2C+last+2+ChromeAndroid+versions%2C+last+2+Safari+versions%2C+last+2+iOS+versions%2C+last+2+Firefox+versions%2C+last+2+Edge+versions)

# License

MIT


CHANGELOG.md
<a name="2.0.0"></a>
# [2.0.0](https://github.com/greymatter-io/gm-ui-components/compare/v1.4.4...v2.0.0) (2020-12-02)

### Breaking Changes

* add inter as a dev dep and set peer deps to major versions only ([#599](https://github.com/greymatter-io/gm-ui-components/issues/599)) ([2f7a8e6](https://github.com/greymatter-io/gm-ui-components/commit/2f7a8e6))
* deprecate nonstandard svg terms in icon API ([#595](https://github.com/greymatter-io/gm-ui-components/issues/595)) ([5e7708f](https://github.com/greymatter-io/gm-ui-components/commit/5e7708f))
* add react-popper and @popperjs/core as peer dependencies ([4af2b73](https://github.com/greymatter-io/gm-ui-components/commit/4af2b73cee4893c596223b9db9ada8f9a2db3ede))
  
### Bug Fixes

* solve cause of buttons having too-solid text color ([#596](https://github.com/greymatter-io/gm-ui-components/issues/596)) ([12d8554](https://github.com/greymatter-io/gm-ui-components/commit/12d8554))
* update footer social links ([#600](https://github.com/greymatter-io/gm-ui-components/issues/600)) ([78d5c45](https://github.com/greymatter-io/gm-ui-components/commit/78d5c45))

### Features

* icons can now be presented in negative ([#533](https://github.com/greymatter-io/gm-ui-components/issues/533)) ([1955398](https://github.com/greymatter-io/gm-ui-components/commit/1955398))
* update svgr ([#612](https://github.com/greymatter-io/gm-ui-components/issues/612)) ([c6db69c](https://github.com/greymatter-io/gm-ui-components/commit/c6db69c))
* better default breadcrumb delimiter (or custom, or none) ([4e426fc](https://github.com/greymatter-io/gm-ui-components/commit/4e426fc87b02c20ef32522fcf05a4fb2fa7e2aca))




<a name="1.4.4"></a>
## [1.4.4](https://github.com/greymatter-io/gm-ui-components/compare/v1.4.3...v1.4.4) (2020-09-23)


### Bug Fixes

* icon displayName ([#590](https://github.com/greymatter-io/gm-ui-components/issues/590)) ([5c9ef2d](https://github.com/greymatter-io/gm-ui-components/commit/5c9ef2d))
* use different inter package ([#589](https://github.com/greymatter-io/gm-ui-components/issues/589)) ([3fcc21d](https://github.com/greymatter-io/gm-ui-components/commit/3fcc21d))
* move inter to peerDependencies ([159ff70](https://github.com/greymatter-io/gm-ui-components/commit/159ff70))



<a name="1.4.3"></a>
## [1.4.3](https://github.com/greymatter-io/gm-ui-components/compare/v1.4.2...v1.4.3) (2020-09-23)

NOTE: this release uses the wrong package for inter fonts, please upgrade to 1.4.4 to take advantage of the new font-family.

### Bug Fixes

* center spinner loading message ([#577](https://github.com/greymatter-io/gm-ui-components/issues/577)) ([67ddd0d](https://github.com/greymatter-io/gm-ui-components/commit/67ddd0d))


### Features

* performance enhancements ([#585](https://github.com/greymatter-io/gm-ui-components/issues/585)) ([150a9f4](https://github.com/greymatter-io/gm-ui-components/commit/150a9f4))
* use Inter for base type ([#574](https://github.com/greymatter-io/gm-ui-components/issues/574)) ([a4eed1a](https://github.com/greymatter-io/gm-ui-components/commit/a4eed1a))



<a name="1.4.2"></a>
## [1.4.2](https://github.com/greymatter-io/gm-ui-components/compare/v1.4.1...v1.4.2) (2020-08-31)


### Features

* added badge component ([#581](https://github.com/greymatter-io/gm-ui-components/issues/581)) ([498ae48](https://github.com/greymatter-io/gm-ui-components/commit/498ae48))



<a name="1.4.1"></a>
## [1.4.1](https://github.com/greymatter-io/gm-ui-components/compare/v1.4.0...v1.4.1) (2020-06-01)

* Added automatic formatting of JS files on git commit
* Removed outdated related projects from README

<a name="1.4.0"></a>
# [1.4.0](https://github.com/greymatter-io/gm-ui-components/compare/v1.3.0...v1.4.0) (2020-04-09)

### Added components

- IconArt
- IconAt
- IconBuilding
- IconBusiness
- IconCalendarEvent
- IconCalendarEventPin
- IconCalendarPlus
- IconCheckcircleIndeterminate
- IconCity
- IconCitySkyline
- IconColumnToggle
- IconCommentBubble
- IconContrast
- IconCurrencyDollar
- IconEquipment
- IconFilePlus
- IconFlag
- IconFolderOpen
- IconGroupCircle
- IconHashtag
- IconHoliday
- IconHome
- IconLanguage
- IconLink
- IconMagnifierCheck
- IconMagnifierX
- IconMailbox
- IconMaterial
- IconMeasurement
- IconNumber
- IconOrdinal
- IconPathwayProcess
- IconPercent
- IconPhone
- IconPlace
- IconScroll
- IconSocialMention
- IconStop
- IconStreetAddress
- IconTag
- IconTimeDigital
- IconTimeHourglass
- IconToggleColumnLeft
- IconToggleColumnRight
- IconToggleInspector
- IconToggleNavigation
- IconTranslation
- IconWeapons
- IconWww

### Features

* add more icons ([#555](https://github.com/greymatter-io/gm-ui-components/issues/555)) ([8d4f1e0](https://github.com/greymatter-io/gm-ui-components/commit/8d4f1e0))
* improve theme background colors ([#565](https://github.com/greymatter-io/gm-ui-components/issues/565)) ([311537d](https://github.com/greymatter-io/gm-ui-components/commit/311537d))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/greymatter-io/gm-ui-components/compare/v1.2.0...v1.3.0) (2020-01-22)

### Removed components

- IconCopyToWorkspace

### Added components

- Status
- ToggleSwitch
- IconGlobe
- IconCloudUp
- IconColumnToggle
- IconEditLines
- IconDash
- IconEar
- IconFileTypeAudio
- IconFileTypeImage
- IconFileTypeVideo
- IconFolderOpen
- IconSelection
- IconSort09
- IconSort90
- IconSortAZ
- IconSortZA
- IconSortAscending
- IconToggleInspector
- IconToggleNavigation

### Bug Fixes

* hide delimiter prop ([#530](https://github.com/greymatter-io/gm-ui-components/issues/530)) ([b902d41](https://github.com/greymatter-io/gm-ui-components/commit/b902d41))
* remove explicit disabled prop from tabwrap ([#537](https://github.com/greymatter-io/gm-ui-components/issues/537)) ([4d8453d](https://github.com/greymatter-io/gm-ui-components/commit/4d8453dc56ba8e9200bb8b191aec129c61d6f77b))

src/style/theme.js
import {
  setLightness,
  transparentize,
  mix,
  saturate,
  darken,
  desaturate,
  lighten,
  getLuminance
} from "polished";

// Offsets = DEFAULT, ONE, TWO, THREE...
// Sizes = MIN, XS, SM, DEFAULT, LG, XL, MAX
// Sizes = LOOSE, DEFAULT, TIGHT
// Strengths (4) = FULL, LIGHT, LIGHTER, LIGHTEST
// Strengths (5, 7) = LOWEST, LOWER, LOW, MED, HIGH, HIGHER, HIGHEST

const themeBase = {
  THEME_NAME: "default",
  COLOR_SCHEME: "light",
  OS_INFLUENCE: "none",
  LIGHTNESS_BIAS: 0.075,
  COLOR_SCHEME_SATURATION_ADJUSTMENT: 0.3,

  OPACITY_FULL: "1",
  OPACITY_LIGHT: "0.7",
  OPACITY_LIGHTER: "0.5",
  OPACITY_LIGHTEST: "0.15",

  LINE_HEIGHT_LOOSE: 1.6,
  LINE_HEIGHT_DEFAULT: 1.4,
  LINE_HEIGHT_TIGHT: 1.2,

  FONT_WEIGHT_DEFAULT: 400,
  FONT_WEIGHT_MEDIUM: 500,
  FONT_WEIGHT_SEMIBOLD: 600,
  FONT_WEIGHT_BOLD: 700,

  CORNER_RADIUS_SHARP: "2px",
  CORNER_RADIUS_INPUT: "4px",
  CORNER_RADIUS_CARD_SM: "4px",
  CORNER_RADIUS_CARD_DEFAULT: "6px",
  CORNER_RADIUS_CARD_LG: "8px",
  CORNER_RADIUS_MAX: "90000px",

  COLOR_BRAND_PRIMARY: "rgb(0, 122, 255)",
  COLOR_BRAND_SECONDARY: "#00b42b",

  COLOR_CONTENT: "#000",
  COLOR_BACKGROUND: "#fff",

  COLOR_INTENT_HIGHLIGHT: "rgb(0, 122, 255)",
  COLOR_INTENT_SUCCESS: "#00b42b",
  COLOR_INTENT_DANGER: "#D83D22", // WCAG AA+
  COLOR_INTENT_WARNING: "#F7CD45", // WCAG AA+
  COLOR_INTENT_INFO: "#1E6DF6", // WCAG AA+

  FONT_STACK_DEFAULT: `Inter var, Inter, -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif`,
  FONT_STACK_BRAND: `'Metropolis', 'Helvetica Neue', Arial, sans-serif`,
  FONT_STACK_CODE: `'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace`,

  FONT_SIZE_MIN: 10,
  FONT_SCALE_FACTOR: 2,

  LETTER_SPACING_DEFAULT: "normal",

  FOCUS_RING_SIZE: "2.5px",
  CONTROL_BORDER_WIDTH: "1px",
  INPUT_BOX_INACTIVE_BORDER_COLOR: "currentColor",

  SPACING_BASE: 8,

  ZINDEX_ABYSS: "-9999",
  ZINDEX_FLOOR: "1",
  ZINDEX_DROPDOWN: "1010",
  ZINDEX_STICKY: "1020",
  ZINDEX_FIXED: "1030",
  ZINDEX_SCRIM: "1040",
  ZINDEX_MODAL: "1050",
  ZINDEX_POPOVER: "1060",
  ZINDEX_TOOLTIP: "1070"
};

const prepareTheme = (theme, content, keyline) => {
  // Take a theme and some optional properties,
  // and construct a full set of theme variables from it.

  const defaultContent = {
    contrast: 1,
    default: 0.85,
    muted: 0.6,
    nonessential: 0.3
  };

  const defaultKeyline = {
    default: 0.08,
    solid: 0.08
  };

  if (!content) {
    content = defaultContent;
  }

  if (!keyline) {
    keyline = defaultKeyline;
  }

  const makeTypeLevel = level => {
    return (
      Math.ceil(level * theme.FONT_SCALE_FACTOR) + theme.FONT_SIZE_MIN + "px"
    );
  };

  // Factor used to increase the mix of the foreground
  // with background, when creating a background level.
  const bgScale = 0.05;

  const makeBackgroundLevel = level => {
    if (theme.COLOR_SCHEME === "dark") {
      return mix(level * bgScale, theme.COLOR_CONTENT, theme.COLOR_BACKGROUND);
    } else {
      return mix(level * bgScale, theme.COLOR_CONTENT, theme.COLOR_BACKGROUND);
    }
  };

  return (theme = {
    ...theme,
    COLOR_CONTENT_CONTRAST: transparentize(
      1 - content.contrast,
      theme.COLOR_CONTENT
    ),
    COLOR_CONTENT_DEFAULT: transparentize(
      1 - content.default,
      theme.COLOR_CONTENT
    ),
    COLOR_CONTENT_MUTED: transparentize(1 - content.muted, theme.COLOR_CONTENT),
    COLOR_CONTENT_NONESSENTIAL: transparentize(
      1 - content.nonessential,
      theme.COLOR_CONTENT
    ),

    COLOR_KEYLINE_DEFAULT: lighten(
      theme.LIGHTNESS_BIAS,
      transparentize(1 - keyline.default, theme.COLOR_CONTENT)
    ),
    // TODO: Verify _SOLID version
    COLOR_KEYLINE_SOLID: lighten(
      theme.LIGHTNESS_BIAS,
      mix(1 - keyline.solid, theme.COLOR_BACKGROUND, theme.COLOR_CONTENT)
    ),

    // TODO: make these values align to the current color scheme
    COLOR_BACKGROUND_DEFAULT: lighten(
      theme.LIGHTNESS_BIAS,
      makeBackgroundLevel(0)
    ),
    COLOR_BACKGROUND_TWO: lighten(theme.LIGHTNESS_BIAS, makeBackgroundLevel(1)),
    COLOR_BACKGROUND_THREE: lighten(
      theme.LIGHTNESS_BIAS,
      makeBackgroundLevel(2)
    ),
    COLOR_BACKGROUND_FOUR: lighten(
      theme.LIGHTNESS_BIAS,
      makeBackgroundLevel(3)
    ),
    COLOR_BACKGROUND_FIVE: lighten(
      theme.LIGHTNESS_BIAS,
      makeBackgroundLevel(4)
    ),

    FONT_SIZE_PAGE_TITLE: makeTypeLevel(22),

    FONT_SIZE_HEADING_LG: makeTypeLevel(16),
    FONT_SIZE_HEADING_DEFAULT: makeTypeLevel(15),
    FONT_SIZE_HEADING_SM: makeTypeLevel(13),

    FONT_SIZE_SUBHEADING_LG: makeTypeLevel(6),
    FONT_SIZE_SUBHEADING_DEFAULT: makeTypeLevel(4),
    FONT_SIZE_SUBHEADING_SM: makeTypeLevel(3),

    FONT_SIZE_ITEM_TITLE_LG: makeTypeLevel(4),
    FONT_SIZE_ITEM_TITLE_DEFAULT: makeTypeLevel(3),
    FONT_SIZE_ITEM_TITLE_SM: makeTypeLevel(2),

    FONT_SIZE_TEXT_XL: makeTypeLevel(5),
    FONT_SIZE_TEXT_LG: makeTypeLevel(4),
    FONT_SIZE_TEXT_DEFAULT: makeTypeLevel(3),
    FONT_SIZE_TEXT_SM: makeTypeLevel(2),
    FONT_SIZE_TEXT_XS: makeTypeLevel(1)

    // TODO: Add alpha-unblended versions of bg colors
  });
};

export const addMacOSXInfluence = theme => ({
  ...theme,
  OS_INFLUENCE: "macOS X",
  // Rounded corners
  CORNER_RADIUS_SHARP: "2px",
  CORNER_RADIUS_INPUT: "4px",
  CORNER_RADIUS_CARD_SM: "5px",
  // Apple color palatte
  COLOR_INTENT_HIGHLIGHT: "rgb(0,122,255)",
  COLOR_INTENT_DANGER: "rgb(255,59,48)",
  COLOR_INTENT_WARNING: "rgb(255,149,0)",
  COLOR_INTENT_INFO: "rgb(90,200,250)",
  COLOR_INTENT_SUCCESS: "rgb(40,205,65)"
});

export const addMacOS11Influence = theme => ({
  ...theme,
  OS_INFLUENCE: "macOS 11",
  // More rounded corners
  CORNER_RADIUS_SHARP: "3px",
  CORNER_RADIUS_INPUT: "5px",
  CORNER_RADIUS_CARD_SM: "7px",
  CORNER_RADIUS_CARD_DEFAULT: "11px",
  CORNER_RADIUS_CARD_LG: "17px",
  CORNER_RADIUS_MAX: "90000px",
  // Broad padding
  SPACING_BASE: 10,
  // Apple color palatte
  COLOR_INTENT_HIGHLIGHT: "rgb(0,122,255)",
  COLOR_INTENT_DANGER: "rgb(255,59,48)",
  COLOR_INTENT_WARNING: "rgb(255,149,0)",
  COLOR_INTENT_INFO: "rgb(90,200,250)",
  COLOR_INTENT_SUCCESS: "rgb(40,205,65)"
});

export const addWin10Influence = theme => ({
  ...theme,
  OS_INFLUENCE: "Win10",
  // Sharp corners
  CORNER_RADIUS_SHARP: "0px",
  CORNER_RADIUS_INPUT: "0px",
  CORNER_RADIUS_CARD_SM: "0px",
  CORNER_RADIUS_CARD_DEFAULT: "0px",
  CORNER_RADIUS_CARD_LG: "0px",
  CORNER_RADIUS_MAX: "0px",
  // Windows color palatte
  COLOR_INTENT_HIGHLIGHT: "#0078d4",
  COLOR_INTENT_SUCCESS: "#107c10",
  COLOR_INTENT_DANGER: "#a80000",
  COLOR_INTENT_WARNING: "#d83b01",
  COLOR_INTENT_INFO: "#0078d4",

  CONTROL_BORDER_WIDTH: "2px"
});

export const setSpacingBase = (theme, size) =>
  size ? { ...theme, SPACING_BASE: size } : { ...theme };

export const addHighContrastInfluence = theme => {
  const content = {
    contrast: 1,
    default: 0.95,
    muted: 0.7,
    nonessential: 0.5
  };

  const keyline = {
    default: 0.2,
    solid: 0.2
  };

  return (theme = prepareTheme(theme, content, keyline));
};

export function setColorScheme(theme, scheme) {
  const adjust = theme.COLOR_SCHEME_SATURATION_ADJUSTMENT;

  const invertLightness = color => {
    return setLightness((getLuminance(color) - 1) * -1, color);
  };

  if (theme.COLOR_SCHEME !== scheme) {
    let adjustments = {};
    if (scheme === "light") {
      adjustments = {
        LETTER_SPACING_DEFAULT: "normal",
        COLOR_INTENT_HIGHLIGHT: saturate(adjust, theme.COLOR_INTENT_HIGHLIGHT),
        COLOR_INTENT_SUCCESS: saturate(adjust, theme.COLOR_INTENT_SUCCESS),
        COLOR_INTENT_WARNING: saturate(adjust, theme.COLOR_INTENT_WARNING),
        COLOR_INTENT_INFO: saturate(adjust, theme.COLOR_INTENT_INFO),
        COLOR_INTENT_DANGER: saturate(adjust, theme.COLOR_INTENT_DANGER)
      };
    } else if (scheme === "dark") {
      adjustments = {
        LETTER_SPACING_DEFAULT: "normal",
        COLOR_INTENT_HIGHLIGHT: desaturate(
          adjust,
          theme.COLOR_INTENT_HIGHLIGHT
        ),
        COLOR_INTENT_SUCCESS: desaturate(adjust, theme.COLOR_INTENT_SUCCESS),
        COLOR_INTENT_WARNING: desaturate(adjust, theme.COLOR_INTENT_WARNING),
        COLOR_INTENT_INFO: desaturate(adjust, theme.COLOR_INTENT_INFO),
        COLOR_INTENT_DANGER: desaturate(adjust, theme.COLOR_INTENT_DANGER)
      };
    }
    theme = {
      ...theme,
      COLOR_SCHEME: scheme,
      COLOR_CONTENT: invertLightness(theme.COLOR_CONTENT),
      COLOR_BACKGROUND: invertLightness(theme.COLOR_BACKGROUND),
      ...adjustments
    };
    return prepareTheme(theme);
  } else {
    return theme;
  }
}

export function invertColorScheme(theme) {
  // Return a new theme with dark and light inverted

  let newScheme;
  if (getLuminance(theme.COLOR_BACKGROUND) > 0.5) {
    // current scheme is light
    newScheme = "dark";
  } else {
    // current scheme is dark
    newScheme = "light";
  }

  return (theme = setColorScheme(theme, newScheme));
}

export const setColorSchemeDark = theme => setColorScheme(theme, "dark");

export const setColorSchemeLight = theme => setColorScheme(theme, "light");

export const keenInfluence = theme => ({
  ...theme,
  THEME_NAME: "Keen",
  COLOR_BRAND_PRIMARY: "#00b42b",
  COLOR_BRAND_SECONDARY: "#00b42b",
  COLOR_INTENT_HIGHLIGHT: "#00b42b"
});

let currentTheme = themeBase;

let prefs = {
  useKeen: true,
  preferredColorScheme: "auto",
  allowOSInfluence: false
};

if (prefs.useKeen) {
  currentTheme = keenInfluence(currentTheme);
}

if (prefs.preferredColorScheme === "dark") {
  currentTheme = setColorSchemeDark(currentTheme);
} else if (prefs.preferredColorScheme === "light") {
  currentTheme = setColorSchemeLight(currentTheme);
}

if (prefs.allowOSInfluence) {
  if (window.navigator.userAgent.includes("Mac OS X 10")) {
    currentTheme = addMacOSXInfluence(currentTheme);
  } else if (window.navigator.userAgent.includes("Mac OS X 11")) {
    currentTheme = addMacOS11Influence(currentTheme);
  } else if (window.navigator.userAgent.includes("Win64")) {
    currentTheme = addWin10Influence(currentTheme);
  }
}

export const keen = keenInfluence(currentTheme);
export const keenDark = setColorSchemeDark(keenInfluence(currentTheme));

export const theme = prepareTheme(currentTheme);
